### PR TITLE
[1.12.x] Update pithos configmap during upgrade

### PR DESCRIPTION
**Pull request by UnderGreen**
This PR implementing update process for `pithos-cfg` configmap to use the new config template if it was changed in
updated release.

### TODO
- [x] Manually test installation
- [x] Manually test upgrades from 1.10.x version
- [ ] Address review comments
- [ ] Forward-port changes to master

### Additional information
`rigging` was bumped to version from 5.5.x branch and it changes a lot of kubernetes files because for some reason
`rigging` 5.6.0 was used with a newer kubernetes version.

### CODE REVIEW

File: internal/pithosctl/pkg/cluster/config.go [hunk 1]
```diff
@@ -19,67 +19,59 @@ package cluster
...
-	// Keys defines list of S3 keys which should be created during bootstrap
-	Keys []AccessKey
 	// CassandraPodSelector defines labels to select cassandra pods
 	CassandraPodSelector string
-	// Bootstrap represents bootstrapping configuration for pithos application
-	Bootstrap *Bootstrap
+	// PithosSecret defines secret name storing S3 keys
+	PithosSecret string
+	// ReplicationFactor defines replication factor for cassandra keyspace
+	// ReplicationFactor is ignored by pithos during upgrade and could be set to any value
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/cluster/config.go:40]: Is this a configuration that can be set externally or is it purely computed as required (as I've seen in the code)?

Also, not sure I understand the statement about caller's being able to set it to any value - does this simply mean that it is ignored during upgrades? Then I would definitely remove this bit as it is confusing.

File: internal/pithosctl/pkg/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -70,6 +70,16 @@ func (c *Client) Pods(selector, namespace string) ([]v1.Pod, error) {
 	return podList.Items, nil
 }
 
+// PithosConfigMap return configmap containing pithos configuration
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/kubernetes/kubernetes.go:73]: Update the comment

File: internal/pithosctl/pkg/pithos/keys.go [hunk 3]
```diff
@@ -31,19 +31,18 @@ const (
 	secretKeyLength = 20
 )
 
-func generateAccessKey(tenant string, master bool) (*cluster.AccessKey, error) {
+func generateAccessKey(tenant string, master bool) (keyName cluster.KeyString, key *cluster.AccessKey, err error) {
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/pithos/keys.go:34]: Can you not set the master attribute at the call site since it looks like it's not always used?
```go
keyName, masterKey, err := generateAccessKey(masterTenantName)
// ...
masterKey.Master = true
```
and the boolean literals are far from readable:
```go
getAccessKey(tenantName, false) // false what?
```

File: internal/pithosctl/pkg/pithos/update.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+)
+
+// Update performs update of pithos application
+func Update(ctx context.Context, config *cluster.Config) error {
+	pithosSecret, err := config.KubeClient.PithosSecret(config.PithosSecret, config.Namespace)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	config.Keystore.Keys = make(map[cluster.KeyString]cluster.AccessKey, 2)
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/pithos/update.go:40]: I don't think it's expected by the caller that this will modify the config value it accepts as parameter - why not initialize it at call site and pass by value?

File: internal/pithosctl/pkg/pithos/update.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+	return nil
+}
+
+func parseSecret(secret *v1.Secret, config *cluster.Config, keyType string) error {
+	keyName := keyType + ".key"
+	secretName := keyType + ".secret"
+	keyValue, exist := secret.Data[keyName]
+	if !exist {
+		return trace.Errorf("secret %v does not contain data with the key %v", config.PithosSecret, keyName)
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/pithos/update.go:60]: `trace.NotFound` here and below

File: internal/pithosctl/pkg/pithos/update.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	secretValue, exist := secret.Data[secretName]
+	if !exist {
+		return trace.Errorf("secret %v does not contain data with the key %v", config.PithosSecret, secretName)
+	}
+
+	accessKey := cluster.AccessKey{
+		Master: false,
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/pithos/update.go:68]: nit: this is implicitly so - no need to set

File: internal/pithosctl/pkg/kubernetes/kubernetes.go [hunk 7]
```diff
@@ -70,6 +70,16 @@ func (c *Client) Pods(selector, namespace string) ([]v1.Pod, error) {
...
+func (c *Client) PithosSecret(secretName, namespace string) (*v1.Secret, error) {
```
Comment (r0mant) [inline internal/pithosctl/pkg/kubernetes/kubernetes.go:74]: There doesn't seem to be anything Pithos-specific about this function, so I'd just call it `Secret` or `GetSecret`.

File: internal/pithosctl/pkg/pithos/update.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+		Tenant: regularTenantName,
+		Secret: cluster.KeyString(secretValue),
+	}
+
+	if keyType == masterKeyType {
+		accessKey.Master = true
+		accessKey.Tenant = masterTenantName
+	}
+	config.Keystore.Keys[cluster.KeyString(keyValue)] = accessKey
```
Comment (r0mant) [inline internal/pithosctl/pkg/pithos/update.go:77]: I would avoid side-effects on the passed in objects. Return `accessKey` instead and assign at the caller site.

File: resources/pithos.yaml [hunk 9]
```diff
@@ -99,7 +124,7 @@ spec:
       product: pithos
       pithos-role: pithos
       component: pithos
-  replicas: 1
+  replicas: 2
```
Comment (r0mant) [inline resources/pithos.yaml:127]: Just curious, why this change?

Comment (UnderGreen) [reply resources/pithos.yaml:127]: Typically 1 is not enough and we don't have anyone one node profile for the app anyway.

**Review by a-palchikov — APPROVED** [review]
